docs/spec/lsm/adr-002.md

```diff
@@ -0,0 +1,362 @@
+    //    Increment total liquid staked
+    //    Increment validator liquid shares
+}
+
+func RedeemTokens() {
+    ...
+    // If delegator is a NOT liquid staking provider 
+    //    Decrement total liquid staked
+    //    Decrement validator liquid shares
+}
+
+func Slash() {
+    ...
+    // Decrement total liquid staked (since total liquid stake is denominated in tokens, not shares)
+}
+```
+
+### Transaction failure cases
+With the liquid staking caps in consideration, there are additional scenarios that should cause a transaction to fail:
+```go
+
+func Delegate() {
+    ...
+    // If delegator is a liquid staking provider
+    //    Fail transaction if delegation exceeds global liquid staking cap
+    //    Fail transaction if delegation exceeds validator liquid staking cap
+    //    Fail transaction if delegation exceeds validator bond cap
+}
+
+func Undelegate() {
+    ...
+    // If the unbonded delegation is a ValidatorBond
+    //    Fail transaction if the reduction in validator bond would cause the
+    //    existing liquid delegation to exceed the cap
+}
+
+func BeginRedelegate() {
+    ...
+    // If the delegation is a ValidatorBond
+    //    Fail transaction if the reduction in validator bond would cause the
+    //    existing liquid delegation to exceed the cap
+
+    // If delegator is a liquid staking provider
+    //    Fail transaction if delegation exceeds global liquid staking cap
+    //    Fail transaction if delegation exceeds validator liquid staking cap
+    //    Fail transaction if delegation exceeds validator bond cap
+}
+
+func TokenizeShares() {
+    ...
+    // If the delegation is a ValidatorBond
+    //    Fail transaction - ValidatorBond's cannot be tokenized
+
+    // If the sender is NOT a liquid staking provider
+    //    Fail transaction if tokenized shares would exceed the global liquid staking cap
+    //    Fail transaction if tokenized shares would exceed the validator liquid staking cap
+    //    Fail transaction if tokenized shares would exceed the validator bond cap
+}
+```
+
+### Tombstoning
+* When a validator is tombstoned, the tombstoned stake can be excluded from the global cap. This requires a small change to the evidence module
+```go
+func HandleEquivocationEvidence() // in evidence keeper
+   ...
+   k.slashingKeeper.Tombstone(ctx, consAddr)
+   k.stakingKeeper.DecreaseTotalLiquidStaked(validator.GetTotalLiquidShares())
+```
+
+### Bootstrapping total liquid stake
+When upgrading to enable the liquid staking module, the total global liquid stake and total liquid validator shares must be determined. This can be done in the upgrade handler by looping through delegation records and including the delegation in the total if the delegator has a 32-length address. This is implemented by the following function:
+```go
+func RefreshTotalLiquidStaked() {
+  // Resets all validator TotalLiquidShares to 0
+  // Loops delegation records
```

Comment:
It would be good to benchmark this iteration as it will affect migration time and thus, increase the downtime of the upgraded chain (e.g., the Cosmos Hub).

Reply:
What's the best way to benchmark this? Numia reported there are 650k Cosmoshub non-ICA addresses with delegations.

Reply:
@tac0turtle any suggestions here?

Reply:
chatted offline - we can do this during the testnet